src/dos/dos_files.cpp

```diff
@@ -859,11 +863,18 @@ Bit8u FCB_Parsename(Bit16u seg,Bit16u offset,Bit8u parser ,char *string, Bit8u *
 		if (!isvalid(toupper(d))) {string += 2; goto savefcb;} //TODO check (for ret value)
 		fcb_name.part.drive[0]=0;
 		hasdrive=true;
-		if (isalpha(d) && Drives[toupper(d)-'A']) { //Floppies under dos always exist, but don't bother with that at this level
-			; //THIS* was here
-		} else ret=0xff;
-		fcb_name.part.drive[0]=DOS_ToUpper(string[0])-'A'+1; //Always do THIS* and continue parsing, just return the right code
-		string+=2;
+
+		// Floppies under dos always exist, but don't bother with that at this level
+		assert(d <= CHAR_MAX);
+		if (isalpha(d) && Drives[drive_index(static_cast<char>(d))]) {
+			; // THIS* was here
+		} else {
+			ret = 0xff;
+		}
```

Comment:
I have no idea what comment in line 870 refers to…

I suggest rewriting these lines as:
```suggestion
		if (!(isalpha(d) && Drives[drive_index(static_cast<char>(d))])) {
			ret = 0xff;
		}
```

Reply:
Fixed!